### PR TITLE
[OptimizeInstructions] Combine some relational ops joined Or/And (Part 5-6)

### PR DESCRIPTION
**Pull request by MaxGraey**
```rust
(i32(x) >= 0) | (i32(y) >= 0)   ==>   i32(x & y) >= 0
(i64(x) >= 0) | (i64(y) >= 0)   ==>   i64(x & y) >= 0

(i32(x) != -1) | (i32(y) != -1)   ==>   i32(x & y) != -1
(i64(x) != -1) | (i64(y) != -1)   ==>   i64(x & y) != -1
```

### CODE REVIEW

**Comment by kripken** [comment]
In the last two PRs we added `preservesOr` and `preservesAnd`. I think there is a lot of value in adding such higher-level functions since then it is easier to understand and extend in the future. And I think we can do something similar here: the property here is turning an or into an and in the operation, that is, `F(x) | F(y)  ===  F(x & y)`. This is still an isomorphism, just with a different operation on the two spaces, just like `ln(x * y)  ===  ln(x) + ln(y)` is an isomorphism. So I suggest adding a function like `flipsOrToAnd` or maybe `isomorphicWithOrAnd` or `orAndIsomorphic`?

**Comment by MaxGraey** [comment]
I don't think it's isomorphism. `ln(x * y) === ln(x) + ln(y` is logarithmic identities (properties) which very specific for log / exp spaces. But we could group such rules under `inversesOr` and `inversesAnd`?

**Comment by MaxGraey** [comment]
@kripken All right. I refactored in the way as we discussed. Could you take a look on recent changes please?

**Comment by MaxGraey** [comment]
friendly ping for review

File: src/passes/OptimizeInstructions.cpp [hunk 1]
```diff
@@ -2573,6 +2573,28 @@ struct OptimizeInstructions
...
+      //   binary(x, c) | binary(y, c)  =>  binary(x & y, c)
+      Binary *bx, *by;
+      Expression *x, *y;
+      Const *cx, *cy;
+      if (matches(curr->left, binary(&bx, any(&x), ival(&cx))) &&
+          matches(curr->right, binary(&by, any(&y), ival(&cy))) &&
+          bx->op == by->op && x->type == y->type && cx->value == cy->value &&
+          inversesOr(bx)) {
+        bx->left = Builder(*getModule())
+                     .makeBinary(Abstract::getBinary(x->type, And), x, y);
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:2594]: Can we reuse `by` here to save the allocation?

File: src/passes/OptimizeInstructions.cpp [hunk 2]
```diff
@@ -2586,10 +2608,8 @@ struct OptimizeInstructions
       Binary *bx, *by;
       Expression *x, *y;
       Const *cx, *cy;
-      if (matches(curr,
-                  binary(OrInt32,
-                         binary(&bx, any(&x), ival(&cx)),
-                         binary(&by, any(&y), ival(&cy)))) &&
+      if (matches(curr->left, binary(&bx, any(&x), ival(&cx))) &&
+          matches(curr->right, binary(&by, any(&y), ival(&cy))) &&
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:2612]: Where did the check for OrInt32 go?

**Review by kripken — COMMENTED** [review]
Sorry for the delay.

Have you fuzzed this?

File: src/passes/OptimizeInstructions.cpp (hunk 2 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:2612]: It was earlier: https://github.com/WebAssembly/binaryen/blob/9b6f92165072d46331533b767bf9ec26c9ae1b2d/src/passes/OptimizeInstructions.cpp#L2549

File: src/passes/OptimizeInstructions.cpp (hunk 1 above)
Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp]: Good idea! Done!

**Comment by MaxGraey** [comment]
I fuzzed in prev compand PR but I could refuzz it again later

**Comment by MaxGraey** [comment]
refuzzed `ITERATION: 11104`

**Review by kripken — APPROVED** [review]

**Pull request merged**